src/operations/extension/reduction_partial_rows.hpp

```diff
@@ -0,0 +1,237 @@
+    in_.adjust_access_displacement();
+    out_.adjust_access_displacement();
+  }
+  SYCL_BLAS_INLINE bool valid_thread(cl::sycl::nd_item<1> ndItem) const {
+    return true;
+  }
+
+  /*!
+   * @brief get_workgroup_cluster. This function is used to find the optimum
+   * number of work groups required to execute each partial reduction step.
+   */
+  SYCL_BLAS_INLINE index_t get_workgroup_cluster() noexcept {
+    return ((rows_ - 1) / local_memory_rows + 1) * group_count_cols_;
+  }
+  /*!
+   * @brief get_num_workgroup_cluster. This function is used to extend the
+   * number of work group clusters, in order to make sure that at least 4
+   * operations are available per work group. The number 4 is used based on
+   * empirical research.
+   */
+  SYCL_BLAS_INLINE index_t
+  get_num_workgroup_cluster(index_t compute_units) noexcept {
+    return ((4 * compute_units - 1) / get_workgroup_cluster() + 1);
+  }
+
+  /*!
+   * @brief Get the nd_range value which has to be used for kernels that
+   *        intend to call GemmPartial::run().
+   */
+  SYCL_BLAS_INLINE cl::sycl::nd_range<1> get_nd_range(
+      index_t compute_units) noexcept {
+    const cl::sycl::range<1> nwg(get_workgroup_cluster() *
+                                 get_num_workgroup_cluster(compute_units));
+    const cl::sycl::range<1> wgs(work_group_size);
+    return cl::sycl::nd_range<1>(nwg * wgs, wgs);
+  }
+
+  template <typename local_memory_t>
+  SYCL_BLAS_INLINE void eval(local_memory_t scratch,
+                             cl::sycl::nd_item<1> id) noexcept {
+    /* reference to the scratch memory */
+    auto scratch_ptr = scratch.localAcc.get_pointer().get();
+
+    /* workgroup id */
+    const index_t group_id = id.get_group(0);
+    /* Local thread id */
+    const index_t local_id = id.get_local_id(0);
+
+    /* Block row and column */
+    const index_t group_col = group_id / group_count_rows_;
+    const index_t group_row = group_id - group_col * group_count_rows_;
+
+    /* Item row and column within a block */
+    const index_t local_col = local_id / work_group_rows;
+    const index_t local_row = local_id - local_col * work_group_rows;
+
+    /* Global position of the first element processed by the thread */
+    const index_t global_row = group_row * local_memory_rows + local_row;
+    const index_t global_col = group_col * local_memory_cols + local_col;
+
+    /* Total number of item cols in all work groups */
+    const index_t total_item_cols = local_memory_cols * group_count_cols_;
+
+    /* Initialize private reduction registers */
+    element_t accumulators[rows_per_item];
+#pragma unroll
+    for (index_t wpr = 0; wpr < rows_per_item; wpr++) {
+      accumulators[wpr] = init_val;
+    }
+
+    /* Sequential reduction level:
+     * Load multiple elements from the global memory, reduce them together and
+     * store them in the local memory */
+    {
+      index_t elem_col = global_col;
+      while (elem_col < cols_) {
```

Comment:
Have you benchmarked replacing this with a for loop? You indicated that other loops got good performance by replacing while with for.

Reply:
It makes absolutely no difference (this kernel is very memory-bound anyway) but I will make the change because a `for` loop is more suited here